### PR TITLE
client: handle context completion during async I/O call when the client is not mounting

### PR DESCRIPTION
**Pull request by dparmar18**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ffilz** [comment]
We should move error cases to returning 0 when onfinish is non-null, and passing the return value in the completion. At least for APIs that were new for nonblocking I/O. Need to think about old APIs.

**Comment by dparmar18** [comment]
> We should move error cases to returning 0 when onfinish is non-null, and passing the return value in the completion. At least for APIs that were new for nonblocking I/O. Need to think about old APIs.

@ffilz for the async io api i.e. `Client::ll_preadv_pwritev`, this PR fixes two things:
1) when a async io api is called with nullptr context; it suffers a recursive deadlock and leads to a crash, therefore assert that the context is not a nullptr before commencing the async io(https://tracker.ceph.com/issues/63633)
2) as you mentioned, the case where the callback is not finished if the async io is performed on an unmounted client; it now returns `0` while the context completion takes places with `ENOTCONN` (https://tracker.ceph.com/issues/63629)

PTAL and do let me know your thoughts about this

File: src/client/Client.cc [hunk 1]
```diff
@@ -15944,6 +15944,7 @@ int64_t Client::ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov,
                                   Context *onfinish, bufferlist *bl,
                                   bool do_fsync, bool syncdataonly)
 {
```
Comment (ffilz) [inline src/client/Client.cc:15946]: Note the function prototype:
  int64_t ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov, int iovcnt,
                            int64_t offset, bool write,
                            Context *onfinish = nullptr,
                            bufferlist *blp = nullptr,
                            bool do_fsync = false, bool syncdataonly = false);

ll_preadv_pwritev is allowed to be called sync with onfinish == nullptr. So we can't make this assertion.

**Review by ffilz — COMMENTED** [review]
Please note when the function allows onfinish to be null indicating a synchronous call.

File: src/client/Client.cc (hunk 1 above)
Comment (mattbenjamin) [reply src/client/Client.cc:15946]: thank you

Comment (dparmar18) [reply src/client/Client.cc:15946]: > Note the function prototype: int64_t ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov, int iovcnt, int64_t offset, bool write, Context *onfinish = nullptr, bufferlist *blp = nullptr, bool do_fsync = false, bool syncdataonly = false);
> 
> ll_preadv_pwritev is allowed to be called sync with onfinish == nullptr. So we can't make this assertion.

`ll_preadv_pwritev` when provided nullptr context leads to a crash, check this out https://tracker.ceph.com/issues/63633 and this is the test case I tested it with https://github.com/ceph/ceph/pull/54435/commits/cc600321071f0e254064e72bce9c7aeaeb81d988

Comment (dparmar18) [reply src/client/Client.cc:15946]: i had a conversation on cephfs slack channel about this https://ceph-storage.slack.com/archives/C04LVQMHM9B/p1700743336859379 thus came to a conclusion to assert the context

File: src/test/client/nonblocking.cc [hunk 2]
```diff
@@ -149,3 +149,62 @@ TEST_F(TestClient, LlreadvLlwritev) {
...
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-io-client-not-mounted"));
+  readfinish.reset(new C_SaferCond("test-nonblocking-readfinish-io-client-not-mounted"));
+
+  int64_t rc;
+  bufferlist bl;
+
+  ASSERT_EQ(client->ll_release(fh), 0);
+  client->unmount();
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(), nullptr);
+  ASSERT_EQ(rc, -CEPHFS_ENOTCONN);
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/test/client/nonblocking.cc]: i think you're looking at an older version, the code is currently this https://github.com/ceph/ceph/pull/54651/commits/9555d9b2012c2bccdabfe0fc7f92e1f8cf9be627#diff-67a20aae67cffa6ae8fad5bdad208d9a73c1868bf52c353002288ea015be65b3R201-R204

File: src/test/client/nonblocking.cc [hunk 3]
```diff
@@ -149,3 +149,62 @@ TEST_F(TestClient, LlreadvLlwritev) {
...
+  rc = writefinish->wait();
+  ASSERT_EQ(rc, -CEPHFS_ENOTCONN);
+
+  rc = client->ll_preadv_pwritev(fh, iov_in, 2, 0, false, readfinish.get(), &bl);
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/test/client/nonblocking.cc:206]: yeah the rc is expected to be 0, https://github.com/ceph/ceph/pull/54651/commits/9555d9b2012c2bccdabfe0fc7f92e1f8cf9be627#diff-67a20aae67cffa6ae8fad5bdad208d9a73c1868bf52c353002288ea015be65b3R206-R209

File: src/client/Client.cc [hunk 4]
```diff
@@ -15945,8 +15945,10 @@ int64_t Client::ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov,
...
     RWRef_t mref_reader(mount_state, CLIENT_MOUNTING);
-    if (!mref_reader.is_state_satisfied())
+    if (!mref_reader.is_state_satisfied()) {
+      onfinish->complete(-CEPHFS_ENOTCONN);
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/client/Client.cc:15950]: https://github.com/ceph/ceph/pull/54651#discussion_r1414336398

**Comment by dparmar18** [comment]
this case is relevant to what i've addressed in https://github.com/ceph/ceph/pull/54690, therefore moving the commits there and closing this one

**Pull request closed without merging**